[ConstraintSystem] Rework overload ranking based on Sendable conformances

xedin opened the pull request:
- Increase score if conformance constraint finds missing conformances
  
  Add a new score kind to denote presence of missing conformances -
  `SK_MissingSynthesizableConformance`.

  The types with missing conformances are allowed to appear in a
  valid solution but such solutions should be ranked lower comparing
  to solutions fewer or without them.

- Specify whether specialization check allows missing conformances or not

  We have two places where `CompareDeclSpecializationRequest` is used:

  - A performance optimization that compares two generic overloads;
  - Solution ranking that checks all of the selected overloads against
    another solution.

  The former can allow missing conformances and shouldn't prevent
  the solver from checking overloads that differ on `Sendable`
  (because there is no information about what is passed as arguments)
  but the latter, since it has a solution, should prefer Sendable
  overloads over non-Sendable ones if possible
  (i.e. `init<T: Sendable>(_: T)` is a subtype of `init<T>(_: T)`).

Resolves: rdar://117227549
<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/main/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Add a new score kind to denote presence of missing conformances

The types with missing conformances are allowed to appear in a
valid solution but such solutions should be ranked lower comparing
to solutions fewer or without them.

xedin committed:
[CSSimplify] Increase score if conformance constraint finds missing conformances

Although the solver is allowed to use types with missing synthesizable
conformances we need to note their presence in the score in order
to rank overloads without such types higher.

xedin committed:
[CSRanking] Specify whether specialization check allows missing conformances or not

We have two places where `CompareDeclSpecializationRequest` is used:

- A performance optimization that compares two generic overloads;
- Solution ranking that checks all of the selected overloads against
  another solution.

The former can allow missing conformances and shouldn't prevent
the solver from checking overloads that differ on `Sendable`
(because there is no information about what is passed as arguments)
but the latter, since it has a solution, should prefer Sendable
overloads over non-Sendable ones if possible
(i.e. `init<T: Sendable>(_: T)` is a subtype of `init<T>(_: T)`).

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test macOS platform

hborla commented on include/swift/Sema/ConstraintSystem.h:
```diff
@@ -988,8 +988,12 @@ enum ScoreKind: unsigned int {
+  /// A type with a missing conformance(s) that has be synthesized
+  /// or diagnosed later, such types are allowed to appear in
+  /// a valid solution.
+  SK_MissingSynthesizableConformance,
```
Should this go before `SK_UnappliedFunction`?

xedin replied:
I think this is lower impact than unapplied function (property with a missing conformance is still better than unapplied function?) and for ranking we’ve got a flag and a check in compare solutions.

xedin replied:
We discussed this offline and came to the conclusion that it's better to move it up and filter first on Sendability.

xedin committed:
[ConstraintSystem] Move `SK_MissingSynthesizableConformance` before `SK_UnappliedFunction`

This would make sure that if property is non-Sendable we'd pick
a method if it's Sendable instead.

xedin commented:
@swift-ci please test
